app/models/manager_refresh/inventory_collection.rb

```diff
@@ -975,7 +975,17 @@ def db_collection_for_comparison_for_complement_of(manager_uuids_set)
 
     def full_collection_for_comparison
       return arel unless arel.nil?
-      parent.send(association)
+      collection = parent.send(association)
```

Comment:
Right, there is a prepared section for this (we should never modify the base code with specifics like this), and we already do this for infra:

https://github.com/Ladas/manageiq/blob/efaa339af78a230eed513d002ce4487ae162a977/app/models/manager_refresh/inventory_collection_default/infra_manager.rb#L315

We should put it to base class: https://github.com/Ladas/manageiq/blob/b3110ffdd8dcabc816e503877422516296e57ea8/app/models/manager_refresh/inventory_collection_default.rb#L3

Then it should work for all cloud/infra vms. And we should have a simple spec, that checks it works. (at least for AWS for now). Also we need to checks specs for all cloud managers, since the Infra guys were reporting something was failing.

-----

It's questionable whether this is needed (we should have some discussion about this) and it can affect performance a lot. The removal/adding of provider is not a standard workflow, by removing a provider, we should expect to loose all data (this was all done for a specific VMware usecase). Part of the workflow will be replaced by proper soft delete and unique indexes.

That being said, we can add it for now and talk about removal once the unique indexes and proper soft delete are in place.

@agrare  ^

Reply:
I'm good with moving that reconnect block up to the base class since currently that is what standard refresh does.  Technically it should reconnect based on uid_ems not ems_ref since once a VM/instance is disconnected the ems_ref uniqueness isn't guaranteed.